:sparkles: Adds test case to verify High Availability

bhargavmad opened the pull request:
**What this PR does / why we need it**:
Steps followed:
 - Create 2 ControlPlane machines in different AZs
 - Create a MachineDeployment in az2 and deploy nginx service
 - Now, Delete the ControlPlane machine in az2
 - Verify if the nginx service is still accessible. This proves
that the ControlPlane machine in az1 is still able to communicate
with the machines in az2

**Remarks**
_Please let me know if this approach really tests HA_

randomvariable commented:
/ok-to-test

randomvariable commented:
Can you rebase this PR?

vincepri commented on test/e2e/aws_test.go:
```diff
@@ -368,8 +370,57 @@ var _ = Describe("functional tests", func() {
+			setup.subnetId = nil
+			createMachineDeployment(setup)
+
+			By("Creating the LB service")
```
This is a little confusing, we should explain that this is a kubernetes service and not an AWS ELB

bhargavmad commented:
Sorry for the delay. I was out of office. 
Rebased
Addressed initial comment/s

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
/retest

vincepri commented:
@bhargavmad Are you able to look into the CI failures?

bhargavmad commented:
/retest

bhargavmad commented:
> @bhargavmad Are you able to look into the CI failures?

Hi @vincepri, these failures are around for a while now. The [periodic e2e job](https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-cluster-api-provider-aws#periodic-e2e-release-0-4) also fails intermittently. Not sure exactly whats causing them. Need some help here. 
cc: @ncdc

vincepri commented:
I mentioned it because it seems that the test added as part of this PR is failing.

`functional tests Cluster should be highly available`
```
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/aws_test.go:412
Timed out after 900.000s.
Expected
    <bool>: false
to be true
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/aws_test.go:1293
```

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
The tests passed on my local setup. This particular case `functional tests Cluster should be highly available` failed couple of times, passed once and the last attempt failed because prow timed out. I created a PR to increase the timeout. I will rerun the test again, once the timeout increases.

detiber commented on test/e2e/aws_test.go:
```diff
@@ -406,6 +408,42 @@ var _ = Describe("functional tests", func() {
+
+			By("Deleting Control Plane Machine in second AZ")
+			deleteMachine(machineName, setup.namespace)
+			time.Sleep(30 * time.Second)
```
Why the 30 second sleep here, would it be better to update `deleteMachine` to contain an `Eventually` that waits until the Machine is fully deleted?

detiber commented on test/e2e/aws_test.go:
```diff
@@ -406,6 +408,42 @@ var _ = Describe("functional tests", func() {
+			createMachineDeployment(setup)
+
+			By("Creating the nginx service exposed as LB")
+			_, ingressHostname := createLBService(metav1.NamespaceDefault, "test-nginx-svc", clusterK8sClient)
```
This service is going to be deployed on one of the Nodes that are backing the MachineDeployment, so testing against this service will not give us signal as to the resiliency of the Control Plane as much as the MachineDeployment.

detiber commented on test/e2e/aws_test.go:
```diff
@@ -406,6 +408,42 @@ var _ = Describe("functional tests", func() {
+			_, ingressHostname := createLBService(metav1.NamespaceDefault, "test-nginx-svc", clusterK8sClient)
+
+			By("Checking if the service is accessible")
+			pingService(ingressHostname)
```
Since we are testing Control Plane availability, it might be better to just test the ability to write to the API Server. The type of resource should not matter much, but as long as we can create a resource we know that the API Server (and etcd) are in a good state.

detiber commented on test/e2e/aws_test.go:
```diff
@@ -406,6 +408,42 @@ var _ = Describe("functional tests", func() {
+			pingService(ingressHostname)
+
+			By("Deleting Control Plane Machine in second AZ")
+			deleteMachine(machineName, setup.namespace)
```
Since there are only 2 Machines in this Control Plane, deleting one of the Machines will lead to loss of quorum for etcd. Instead we should test against a 3 Machine Control Plane, where the etcd cluster would be degraded but still have quorum.

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
This is really weird. All the tests passed after increasing timeouts on some of the functions. But git reports it as failure. Please look into [this](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/1468/pull-cluster-api-provider-aws-e2e/1220653537466257408)

ncdc commented:
/test pull-cluster-api-provider-aws-e2e

ncdc commented:
@bhargavmad I looked at the e2e runs from this PR, as well as the periodic e2e runs, and I'm not seeing any real consistency in the test failures across the jobs. I'm guessing your most recent failure was the error in the log from ci-e2e.sh line 80, where it failed to kill the heartbeat process. We can keep trying to see how things go.

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
The tests are very flaky, one or the other test keeps failing every time and i cant seem to find the solution. I thought increasing timeouts will resolve the issue, but it is still failing. Will really appreciate some help here.

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

detiber commented on test/e2e/aws_test.go:
```diff
@@ -431,8 +430,64 @@ var _ = Describe("functional tests", func() {
+	})
 })
 
+func createDummyResource(name, namespace string, k8sClient crclient.Client) {
```
Instead of creating a Pod here, I would probably just create a Namespace to test the functionality of the API server.

detiber commented on test/e2e/aws_test.go:
```diff
@@ -431,8 +430,64 @@ var _ = Describe("functional tests", func() {
+			By("Creating the MachineDeployment in second AZ")
+			setup.availabilityZone = availabilityZones[1]
+			setup.subnetId = nil
+			createMachineDeployment(setup)
```
Since you only need to test the availability of the API server here, you can avoid creating a MachineDeployment here.

detiber commented on test/e2e/aws_test.go:
```diff
@@ -790,8 +845,15 @@ func createClusterKubeConfigs(tmpDir, namespace, clusterName string) (string, cr
+			return err
+		},
+		5*time.Minute, 10*time.Second,
+	).Should(BeNil())
```
Since you are testing the result of an error, I would say it is better to use `gomega.Succeed()` here.
```suggestion
	).Should(Succeed())
```

detiber commented on test/e2e/aws_test.go:
```diff
@@ -1130,12 +1210,17 @@ func makeAWSMachineTemplate(namespace, name, instanceType string, az, subnetId *
+			).Run()
+		},
+		5*time.Minute, 10*time.Second,
+	).Should(BeNil())
```
```suggestion
	).Should(Succeed())
```

detiber replied:
I'm a bit curious as the why we would need to wrap this with an Eventually.

bhargavmad replied:
I have observed many failures in the test-grid while trying to deploy CNI. It could be because the DNS for the API server may not be available.

detiber commented:
It looks like the most recent test failure was related to attempting to delete the Cluster following the HA tests added here.

I'll attempt to dig into the logs a bit later to see if I can tell why it is taking longer than 20 minutes to delete.

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
The latest run fails with log
```
I0130 12:03:33.938758       1 awsmachine_controller.go:264] controllers/AWSMachine "msg"="Error state detected, skipping reconciliation" "awsCluster"="test-infra-l49ppd" "awsMachine"="test-infra-capa-mtft3bkv-6kd27" "cluster"="test-fm0hip" "machine"="test-capa-mdyo14ub-787b7568f7-7qv87" "namespace"="test-qrl6my" 
I0130 12:03:59.958947       1 leaderelection.go:281] failed to renew lease capa-system/controller-leader-election-capa: failed to tryAcquireOrRenew context deadline exceeded
E0130 12:03:59.959119       1 main.go:176] setup "msg"="problem running manager" "error"="leader election lost"  
```
It is observed in both capa and capi log files. It failed during Delete cluster step of one of the tests and the subsequent specs fail at `BeforeEach` with the same error ([logs](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/1468/pull-cluster-api-provider-aws-e2e/1222802920391577600/artifacts))

cc: @ncdc @detiber

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
/test pull-cluster-api-provider-aws-e2e

bhargavmad commented:
@detiber I have addressed the review comments and the latest run succeeded except for the error in the `ci-e2e.sh` script. Please do have a look and merge if everything looks good.

detiber commented on test/e2e/aws_test.go:
```diff
@@ -975,13 +1013,31 @@ func deleteMachine(namespace, machineDeploymentName string) {
+				if apierrors.IsNotFound(err) {
+					return true
+				}
+				return false
```
This can be simplified to `return apierrors.IsNotFound(err)`

detiber commented:
/lgtm
/approve
